skyportal/handlers/api/internal/image_analysis.py

```diff
@@ -0,0 +1,607 @@
+    ra2: float
+        Second set of points RA
+    dec2: float
+        Second set of points Dec
+    sr:
+        Maximal acceptable pair distance to be considered a match, in degrees
+
+    Returns
+    -------
+    tuple
+        Two parallel sets of indices corresponding to matches from first and second lists, along with the pairwise distances in degrees
+
+    """
+    try:
+        idx1, idx2, dist, _ = search_around_sky(
+            SkyCoord(ra1, dec1, unit='deg'), SkyCoord(ra2, dec2, unit='deg'), sr * u.deg
+        )
+
+        dist = dist.deg  # convert to degrees
+
+        return idx1, idx2, dist
+    except Exception as e:
+        raise e
+
+
+def spherical_distance(ra1, dec1, ra2, dec2):
+    """Compute the distance between two points on the sphere.
+
+    Parameters
+    ----------
+    ra1: float
+        First point RA
+    dec1: float
+        First point Dec
+    ra2: float
+        Second point RA
+    dec2: float
+        Second point Dec
+
+    Returns
+    -------
+    float
+        Distance in degrees
+
+    """
+    try:
+        x = np.sin(np.deg2rad((ra1 - ra2) / 2))
+        x *= x
+        y = np.sin(np.deg2rad((dec1 - dec2) / 2))
+        y *= y
+
+        z = np.cos(np.deg2rad((dec1 + dec2) / 2))
+        z *= z
+
+        return np.rad2deg(2 * np.arcsin(np.sqrt(x * (z - y) + y)))
+    except Exception as e:
+        raise e
+
+
+def reduce_image(
+    image,
+    header,
+    obj_id,
+    instrument_id,
+    user_id,
+    gain,
+    matching_radius,
+    catalog_name_refinement,
+    catalog_name_crossmatch,
+    template_name,
+    method,
+    detect_cosmics=False,
+):
+    """Reduce an image: Perform astrometric and photometric calibration, and extract photometry to add it to the database.
+
+    Parameters
+    ----------
+    image: numpy.ndarray
+        Image data
+    header: astropy.io.fits.Header
+        Image header
+    obj_id: str
+        Object ID
+    instrument_id: int
+        Instrument ID
+    user_id: int
+        User ID
+    detect_cosmics: bool
+        Run LACosmic on the image
+
+    Returns
+    -------
+    None
+
+    """
+
+    try:
+        workdir_sextractor_obj = tempfile.mkdtemp(prefix='sex')
+        workdir_sextractor_obj1 = tempfile.mkdtemp(prefix='sex1')
+        workdir_psfex = tempfile.mkdtemp(prefix='psfex')
+
+        with DBSession() as session:
+            instrument = session.scalars(
+                sa.select(Instrument).where(Instrument.id == instrument_id)
+            ).first()
+            user = session.scalars(sa.select(User).where(User.id == user_id)).first()
+
+            time = utils.get_obs_time(header, verbose=False)
+            filt = header.get('FILTER')
+
+            wcs = WCS(header)
+
+            # Masking saturated stars and cosmic rays
+            # The mask is a binary frame with the same size as the image where True means that this pixel should not be used for the analysis
+            mask = image > 0.9 * np.max(image)
```

Comment:
I would guess you would also want to use "dilate" on this mask before applying it, to get the edges that are not saturated.

Reply:
I am not familiar with that process (especially with NumPy), do have any examples in mind? And also, what would you use to do so, a library like openCV?

Reply:
I have found someone doing something fairly similar at: https://github.com/stubbslab/general_purpose_python_scripts/blob/main/cosmics.py

For instance, they define a method that does:
```
def getdilatedmask(self, size=3):
		"""
		Returns a morphologically dilated copy of the current mask.
		size = 3 or 5 decides how to dilate.
		"""
		if size == 3:
			dilmask = ndimage.morphology.binary_dilation(self.mask, structure=growkernel, iterations=1, mask=None, output=None, border_value=0, origin=0, brute_force=False)
		elif size == 5:
			dilmask = ndimage.morphology.binary_dilation(self.mask, structure=dilstruct, iterations=1, mask=None, output=None, border_value=0, origin=0, brute_force=False)
		else:
			dismask = self.mask.copy()
		
		return dilmask
```
so using scipy

Reply:
Hi @mcoughlin @guynir42. Any thoughts about this?

Reply:
I think we can address this in a follow-on PR.